ddsrouter_utils/include/ddsrouter_utils/pool/LimitlessPool.hpp

```diff
@@ -0,0 +1,129 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @file LimitlessPool.hpp
+ */
+
+#ifndef _DDSROUTERUTILS_POOL_LIMITLESSPOOL_HPP_
+#define _DDSROUTERUTILS_POOL_LIMITLESSPOOL_HPP_
+
+#include <queue>
+
+#include <ddsrouter_utils/pool/IPool.hpp>
+
+namespace eprosima {
+namespace ddsrouter {
+namespace utils {
+
+/**
+ * @brief This class implements a generic not-thread-safe reusable Pool without size limit.
+ *
+ * ATTRIBUTES:
+ * - Reuse freed elements without allocation.
+ * - Not thread sage
+ * - Not pool limit
+ *
+ * This implementation uses an internal queue where new elements are stored.
+ * Whenever it runs out of free (not loaned) elements, it allocates more following configuration batch.
+ * Whenever an element is returned, it turns back to the queue to be reused.
+ *
+ * @warning This class is not thread safe.
+ *
+ * @tparam T Type of the elements in the pool.
+ */
+template <typename T>
+class LimitlessPool : IPool<T>
+{
+public:
+
+    /**
+     * @brief Create a new LimitlessPool object by a Pool Configuration given.
+     *
+     * It initializes the internal vector with the given initial size, using \c new_element_ .
+     *
+     * @note Whenever this class is inherited, the constructor must call \c initialize_vector_ ,
+     * as this class cannot because \c new_element_ must be override by the child class.
+     *
+     * @param configuration Pool Configuration
+     */
+    LimitlessPool(
+            PoolConfiguration configuration);
+
+    /**
+     * @brief Destroy the Limitless Pool object and all its reserved elements.
+     */
+    ~LimitlessPool();
+
+    /**
+     * @brief Override IPool::loan
+     *
+     * If there are elements not in use in the vector, it returns one of them in \c element .
+     * Otherwise, it allocates more elements and take the next free one.
+     */
+    virtual bool loan(
+            T*& element) override;
+
+    /**
+     * @brief Override IPool::return_loan
+     *
+     * Return the element to the vector.
+     *
+     * @throw InconsistencyException if there have been more released than reserved calls.
+     */
+    virtual bool return_loan(
+            T* element) override;
+
+protected:
+
+    /**
+     * @brief Initialize the internal queue with the given initial size in configuration.
+     *
+     * This method must be called in every final child class constructor.
+     */
+    virtual void initialize_queue_();
+
+    /**
+     * @brief Augment the internal queue with \c batch new elements using \c new_element_ function for each of them.
+     */
+    void augment_free_values_();
+
+    /**
+     * @brief queue where elements are stored.
+     *
+     * Each element of the queue has been initialized with \c new_element_
+     * or returned to the pool after calling \c reset_element_ .
+     * The elements are consumed in queue order.
+     */
+    std::queue<T*> elements_;
```

Comment:
This depends also on the size of the collection and the multithreading access to it.
If multithreading is expected (this is not the case, but could be in the future) queue is faster. (This was the main reason why I use queue, but it is true that it is not required here).
Also, If elements are not meant to be returned (as in out specific case) vector is less efficient because it would only increase, and increase vectors means to REallocate the whole vector somewhere else, with copies included. While `deque` based objects allocate different chunks of data, so it would be faster.

Summary: I am not sure which is the best one, that is why I wanted to do it generic and create one or the other depending on the scenario. But I agree vector could be use in this example and would not be much difference.

Reply:
I am not sure if the reallocation issue would only occur with random insertions. Anyways, perhaps you wanna have a look at this nice post: https://baptiste-wicht.com/posts/2012/12/cpp-benchmark-vector-list-deque.html